### PR TITLE
feat: support namespaced functions

### PR DESCRIPTION
**Pull request by ansgarm**
🚫 blocked until there's a new release of `hcl/release` with support for `ExprSyntaxError`:
https://github.com/hashicorp/hcl-lang/blob/905e5636294807904a67f6c89a0562db0b82f32c/go.mod#L8

Closes https://github.com/hashicorp/vscode-terraform/issues/1697

### CODE REVIEW

File: schema/signature.go
Comment (ansgarm) [inline schema/signature.go:39]: @dbanck, do you think this needs a deep(er) copy?

Comment (dbanck) [reply schema/signature.go:39]: Yes, we usually need to make sure to explicitly copy slicing and pointing.

File: schema/signature.go
Comment (dbanck) [inline schema/signature.go:36]: Instead of using a helper function here, that reimplements some of the `cty` internals (field names) and would need to be updated whenever something upstream changes, we can make a copy using some pointer magic.

```go
if fs.VarParam != nil {
	vpCpy := *fs.VarParam
	newFS.VarParam = &vpCpy
}
```

Comment (ansgarm) [reply schema/signature.go]: Ah, neat! And that works because there are no nested pointers, right? Or would that be irrelevant anyway?

Comment (dbanck) [reply schema/signature.go]: While it possible to have nested pointers in Go (`**T`), I don't think it's relevant here.

File: schema/signature.go
Comment (dbanck) [inline schema/signature.go:38]:
```suggestion
        newFS.Params = make([]function.Parameter, len(fs.Params))
	copy(newFS.Params, fs.Params)
```

We still need to create the slice before copying something into it.

File: decoder/expr_function.go
Comment (dbanck) [inline decoder/expr_function.go:300]:
```suggestion
func hoverContentForFunction(name string, funcSig schema.FunctionSignature) lang.MarkupContent {
```

Since we are using only one field (`name)`, we could pass it to the function as a string instead of passing the whole expression.

File: go.mod
Comment (dbanck) [inline go.mod:8]: I think we should wait for an HCL release before merging this

**Review by dbanck — CHANGES_REQUESTED** [review]
Nice work, some minor suggestions.

Thanks for adding a namespaced function to the tests

File: schema/signature.go
Comment (ansgarm) [reply schema/signature.go:39]: Well that makes sense 😅

File: decoder/expr_function.go
Comment (ansgarm) [inline decoder/expr_function.go:61]: Is this too edge-casey? Do we need a follow-up issue for this?

File: decoder/expr_function.go
Comment (ansgarm) [inline decoder/expr_function.go:85]: We are currently only looking for the `provider:` prefix, but in theory `hcl-lang` should support any namespaced function, right?

Comment (dbanck) [reply decoder/expr_function.go:85]: Yes, we should avoid implementing anything Terraform specific in `hcl-lang`. Any namespaced function should work and also the short syntax `namespace::func()`

File: decoder/expr_function.go
Comment (dbanck) [reply decoder/expr_function.go:61]: If you want, you could do a quick check to see if you can catch the `ScopeTraversalExpr` case, and if it gets too complicated, create a follow-on issue. But I'm fine with creating a follow-on issue right away.

Comment (ansgarm) [reply decoder/expr_function.go]: Created a follow-up issue: https://github.com/hashicorp/hcl-lang/issues/385

**Review by dbanck — APPROVED** [review]

**Pull request merged**
